File: 384.java
// Leetcode 384 Shuffle an Array
// Reference: https://leetcode.com/problems/shuffle-an-array/solutions/85958/first-accepted-solution-java/?envType=featured-list&envId=top-interview-questions?envType=featured-list&envId=top-interview-questions

import java.util.Random;

class Solution {
    private int[] nums;
    private Random random;

    public Solution(int[] nums) {
        this.nums = nums;
        random = new Random();

        
    }
    
    public int[] reset() {

        return nums;
        
    }
    
    public int[] shuffle() {
        if(nums == null) return null;

        int[] a = nums.clone();
        for (int j = 1; j < a.length;j++){
            int i = random.nextInt(j+1);
            swap(a, i, j);
        }
        return a;
        
    }

    private void swap(int[]a, int i, int j){
        int t = a[i];
        a[i] = a[j];
        a[j] = t;
    }
}

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(nums);
 * int[] param_1 = obj.reset();
 * int[] param_2 = obj.shuffle();
 */

File: 384.js
// Leetcode 384 Shuffle an Array
// Reference:  https://leetcode.com/problems/shuffle-an-array/?envType=featured-list&envId=top-interview-questions?envType=featured-list&envId=top-interview-questions

/**
 * @param {number[]} nums
 */
var Solution = function(nums) {
    this.nums = nums;
    this.original = nums.slice();
};

/**
 * @return {number[]}
 */
Solution.prototype.reset = function() {
    return this.original;
    
};

/**
 * @return {number[]}
 */
Solution.prototype.shuffle = function() {
    this.nums.sort((a,b) => {
        if(Math.random() > 0.5) return 1;
        else return -1;
    })

    return this.nums;
    
};

/** 
 * Your Solution object will be instantiated and called as such:
 * var obj = new Solution(nums)
 * var param_1 = obj.reset()
 * var param_2 = obj.shuffle()
 */